language/diem-framework/modules/Diem.move

```diff
@@ -485,97 +551,338 @@ module Diem {
+        assert(
+            Vector::length(preburns) < MAX_OUTSTANDING_PREBURNS,
+            Errors::limit_exceeded(EPREBURN_QUEUE)
+        );
+        Vector::push_back(preburns, preburn);
+    }
+    spec fun add_preburn_to_queue {
+        pragma opaque;
+        let account_addr = Signer::spec_address_of(account);
+        let preburns = global<PreburnQueue<CoinType>>(account_addr).preburns;
+        modifies global<PreburnQueue<CoinType>>(account_addr);
+        aborts_if !exists<PreburnQueue<CoinType>>(account_addr) with Errors::INVALID_STATE;
+        include AddPreburnToQueueAbortsIf<CoinType>;
+        ensures exists<PreburnQueue<CoinType>>(account_addr) ==>
+            Vector::eq_push_back(preburns, old(preburns), preburn);
+        ensures !exists<PreburnQueue<CoinType>>(account_addr) ==>
+            preburns == Vector::spec_singleton(preburn);
+    }
+    spec schema AddPreburnToQueueAbortsIf<CoinType> {
+        account: signer;
+        preburn: Preburn<CoinType>;
+        let account_addr = Signer::spec_address_of(account);
+        aborts_if preburn.to_burn.value == 0 with Errors::INVALID_ARGUMENT;
+        aborts_if exists<PreburnQueue<CoinType>>(account_addr) &&
+            Vector::length(global<PreburnQueue<CoinType>>(account_addr).preburns) >= MAX_OUTSTANDING_PREBURNS
+            with Errors::LIMIT_EXCEEDED;
+    }
+
     /// Sends `coin` to the preburn queue for `account`, where it will wait to either be burned
     /// or returned to the balance of `account`.
-    /// Calls to this function will fail if `account` does not have a
-    /// `Preburn<CoinType>` resource published under it.
+    /// Calls to this function will fail if:
+    /// * `account` does not have a `PreburnQueue<CoinType>` resource published under it; or
+    /// * the preburn queue is already at capacity (i.e., at `MAX_OUTSTANDING_PREBURNS`); or
+    /// * `coin` has a `value` field of zero.
     public fun preburn_to<CoinType>(
         account: &signer,
         coin: Diem<CoinType>
-    ) acquires CurrencyInfo, Preburn {
+    ) acquires CurrencyInfo, Preburn, PreburnQueue {
         Roles::assert_designated_dealer(account);
+        // any coin that is preburned needs to have a nonzero value
+        assert(value(&coin) > 0, Errors::invalid_argument(ECOIN));
         let sender = Signer::address_of(account);
-        assert(exists<Preburn<CoinType>>(sender), Errors::not_published(EPREBURN));
-        preburn_with_resource(coin, borrow_global_mut<Preburn<CoinType>>(sender), sender);
+        // After an upgrade a `Preburn` resource no longer exists in this
+        // currency, and it is replaced with a `PreburnQueue` resource
+        // for the same currency.
+        upgrade_preburn<CoinType>(account);
+
+        let preburn = Preburn {
+            to_burn: zero<CoinType>(),
+        };
+        preburn_with_resource(coin, &mut preburn, sender);
+        add_preburn_to_queue(account, preburn);
```

Comment:
It might be nice to combine this with `preburn_with_resource` (since you would never want to do one without the other).

Reply:
Just had a look at what would happen if we tried to combine these...sadly if we combine them, then we need to change the logic in `burn_now`, which then causes a number of rippled effects out to the transaction fees as well. So I would have a preference for keeping this as-is, but interested in your take